src/main/java/fi/aalto/cs/apluscourses/intellij/actions/TutorialAction.java

```diff
@@ -87,13 +88,16 @@ private void doTutorial(@NotNull Project project) {
     Optional.ofNullable(mainViewModel.tutorialViewModel.get())
         .ifPresent(TutorialViewModel::cancelTutorial);
 
+    TaskNotifier taskNotifier = new TaskNotifier(notifier, project);
+
     TutorialViewModel tutorialViewModel =
         new TutorialViewModel(tutorialExercise, new IntelliJActivityFactory(project),
-            new TaskNotifier(notifier, project), dialogs);
+            taskNotifier, dialogs);
     if (dialogs.confirmStart(tutorialViewModel)) {
+      tutorialViewModel.getTutorial().downloadDependencies(courseViewModel.getModel(), project, taskNotifier);
       mainViewModelProvider.getMainViewModel(project).tutorialViewModel.set(tutorialViewModel);
       tutorialViewModel.getTutorial().tutorialCompleted
-          .addListener(mainViewModel, this::onTutorialComplete);
+          .addListener(mainViewModel, mainVm -> onTutorialComplete(mainVm, courseViewModel.getModel()));
```

Comment:
`Event`s (and `ObservableProperty`s) were designed to handle data flow from model to view. That's why they use weak references. The idea was that the view gets updated as long as it exists and when it no longer exists, it will be GC'd even if the model was still there. Another safe use case of these is data flow from "child objects" to "parent objects" (like from `Module` to `Course`), when parent already holds references to its children. If a child gets removed from a parent, it will get eventually GC'd (though the parent can still get updates from it before GC steps in).

To me, it seems this use case doesn't fit in either of the above-mentioned, so perhaps using our `Event` class is not the best idea here... 🤔 In practice, the weak referencing probably doesn't matter here, though, because the main view model is probably alive longer than the tutorial. But it is difficult to see here whether it's correct or not. The lambda creating a closure with `this` (an action object) and `courseViewModel` makes it even more complicated.

Perhaps just a crude observer-implementation without weak referencing would make more sense. We can discuss it, if you like.

Reply:
This concern still holds.